Make CI image signing reliable

I've seen more than an acceptable level of CI image signing errors during CI builds recently. We should investigate how to make this step more reliable, or maybe optionally skip it for CI builds if it is introducing too much load on the underlying service.

Example failure on a PR: https://github.com/cilium/cilium/actions/runs/16943844870/job/48019697243?pr=41135#step:24:1

```
cosign sign -y quay.io/cilium/docker-plugin-ci@sha256:c7a1b5dbb2b2b0f18aa7721eabc75d8fb7aa1577d4afc3ba2932cb28ddca0320
  if [[ "success" != 'skipped' ]]; then
    cosign sign -y quay.io/cilium/docker-plugin-ci@sha256:a1a2114fc96f111ba082c41626bad26e85ef7b7840fbc7faab6459bb47b18c2f
  fi
  if [[ "success" != 'skipped' ]]; then
    cosign sign -y quay.io/cilium/docker-plugin-ci@sha256:c87e23cb15a6d8528606503f2d514895ba4dad61f0da22a76709a41d256fdc25
  fi
  shell: /usr/bin/bash -e {0}
  env:
    QUAY_ORGANIZATION: cilium
    QUAY_ORGANIZATION_DEV: cilium
    QUAY_CHARTS_ORGANIZATION_DEV: cilium-charts-dev
    EGRESS_GATEWAY_HELM_VALUES: --helm-set=egressGateway.enabled=true
    BGP_CONTROL_PLANE_HELM_VALUES: --helm-set=bgpControlPlane.enabled=true
    CILIUM_CLI_RELEASE_REPO: cilium/cilium-cli
    CILIUM_CLI_VERSION: 
    CILIUM_CLI_IMAGE_REPO: quay.io/cilium/cilium-cli-ci
    CILIUM_CLI_SKIP_BUILD: true
    CILIUM_CLI_CODE_OWNERS_PATHS: CODEOWNERS
    CILIUM_CLI_EXCLUDE_OWNERS: @cilium/github-sec
    PUSH_TO_DOCKER_HUB: true
    GCP_PERF_RESULTS_BUCKET: gs://cilium-scale-results
    CILIUM_RUNTIME_IMAGE_PREFIX: quay.io/cilium/
    KIND_VERSION: v0.29.0
    KIND_K8S_IMAGE: quay.io/cilium/kindest-node:v1.34.0-rc.1@sha256:c60d9e8ae62dfa5128537407005efd1ae7965673b200f4da2c362bc0998a0322
    KIND_K8S_VERSION: v1.34.0-rc.1
    CILIUM_RUNTIME_IMAGE: quay.io/cilium/cilium-runtime:faaac4f040a5b14270b7122c084aca19c370776c@sha256:e01e27f4c241eb37cc1fbc5e07ef9927f89a8a4cc8e06b61e70831997652aaf3
setting TUF refresh period to 24h0m0s
Generating ephemeral keys...
Retrieving signed certificate...
Successfully verified SCT...

	The sigstore service, hosted by sigstore a Series of LF Projects, LLC, is provided pursuant to the Hosted Project Tools Terms of Use, available at [https://lfprojects.org/policies/hosted-project-tools-terms-of-use/.](https://lfprojects.org/policies/hosted-project-tools-terms-of-use/)
	Note that if your submission includes personal data associated with this signed artifact, it will be part of an immutable record.
	This may include the email address associated with the account with which you authenticate your contractual Agreement.
	This information will be used for signing this artifact and will be stored in public transparency logs and cannot be removed later, and is subject to the Immutable Record notice at [https://lfprojects.org/policies/hosted-project-tools-immutable-records/.](https://lfprojects.org/policies/hosted-project-tools-immutable-records/)

By typing 'y', you attest that (1) you are not submitting the personal data of any other person; and (2) you understand and agree to the statement and the Agreement terms at the URLs listed above.
tlog entry created with index: 390237817
Pushing signature to: quay.io/cilium/docker-plugin-ci
setting TUF refresh period to 24h0m0s
Generating ephemeral keys...
Retrieving signed certificate...
Error: signing [quay.io/cilium/docker-plugin-ci@sha256:a1a2114fc96f111ba082c41626bad26e85ef7b7840fbc7faab6459bb47b18c2f]: getting signer: getting key from Fulcio: retrieving cert: client: Post "https://fulcio.sigstore.dev/api/v1/signingCert": dial tcp 34.36.164.164:443: i/o timeout
error during command execution: signing [quay.io/cilium/docker-plugin-ci@sha256:a1a2114fc96f111ba082c41626bad26e85ef7b7840fbc7faab6459bb47b18c2f]: getting signer: getting key from Fulcio: retrieving cert: client: Post "https://fulcio.sigstore.dev/api/v1/signingCert": dial tcp 34.36.164.164:443: i/o timeout
Error: Process completed with exit code 1.
```

--- a/.github/actions/cosign/action.yaml
+++ b/.github/actions/cosign/action.yaml
@@ -22,6 +22,14 @@ inputs:
     description: 'Upload SBOM as a release asset'
     required: false
     default: 'true'
+  max_attempts:
+    description: 'Maximum number of retry attempts'
+    required: false
+    default: '5'
+  initial_delay:
+    description: 'Initial delay in seconds before first retry'
+    required: false
+    default: '10'
 
 runs:
   using: 'composite'
@@ -42,11 +50,62 @@ runs:
     - name: Sign Container Image
       shell: bash
       run: |
-        cosign sign -y "${{ inputs.image }}"
+        # Function to retry cosign signing with exponential backoff
+        sign_with_retry() {
+          local image="$1"
+          local max_attempts=${{ inputs.max_attempts }}
+          local delay=${{ inputs.initial_delay }}
+
+          for attempt in $(seq 1 $max_attempts); do
+            echo "Attempt $attempt/$max_attempts: Signing $image"
+            if cosign sign -y "$image"; then
+              echo "Successfully signed $image"
+              return 0
+            else
+              echo "Failed to sign $image (attempt $attempt/$max_attempts)"
+              if [ $attempt -lt $max_attempts ]; then
+                echo "Waiting ${delay}s before retry..."
+                sleep $delay
+                delay=$((delay * 2))  # Exponential backoff
+              fi
+            fi
+          done
+
+          echo "ERROR: Failed to sign $image after $max_attempts attempts"
+          return 1
+        }
+
+        sign_with_retry "${{ inputs.image }}"
 
     - name: Attach SBOM Attestation
       shell: bash
       run: |
+        # Function to retry cosign attestation with exponential backoff
+        attest_with_retry() {
+          local predicate="$1"
+          local image="$2"
+          local max_attempts=${{ inputs.max_attempts }}
+          local delay=${{ inputs.initial_delay }}
+
+          for attempt in $(seq 1 $max_attempts); do
+            echo "Attempt $attempt/$max_attempts: Attesting $image with $predicate"
+            if cosign attest -y --predicate "$predicate" --type spdxjson "$image"; then
+              echo "Successfully attested $image"
+              return 0
+            else
+              echo "Failed to attest $image (attempt $attempt/$max_attempts)"
+              if [ $attempt -lt $max_attempts ]; then
+                echo "Waiting ${delay}s before retry..."
+                sleep $delay
+                delay=$((delay * 2))  # Exponential backoff
+              fi
+            fi
+          done
+
+          echo "ERROR: Failed to attest $image after $max_attempts attempts"
+          return 1
+        }
+
         image="${{ inputs.image }}"
         sbom_file="./sbom_${{ inputs.sbom_name }}.spdx.json"
-        cosign attest -y --predicate "${sbom_file}" --type spdxjson "${image}"
+        attest_with_retry "${sbom_file}" "${image}"

--- a/.github/actions/cosign/action.yaml
+++ b/.github/actions/cosign/action.yaml
@@ -1,0 +1,52 @@
+name: 'Generate SBOM and Sign with Cosign'
+description: 'Generate SBOM, sign container images, and attach SBOM attestations'
+inputs:
+  image:
+    description: 'Container image to sign (format: "quay.io/org/image@digest")'
+    required: true
+  image_tag:
+    description: 'Container image tag for SBOM generation (format: "quay.io/org/image:tag")'
+    required: false
+  sbom_name:
+    description: 'Name for the SBOM artifact and file (e.g., "cilium_v1.2.3")'
+    required: true
+  generate_sbom:
+    description: 'Generate SBOM'
+    required: false
+    default: 'true'
+  install_cosign:
+    description: 'Whether to install cosign (set to false if already installed)'
+    required: false
+    default: 'true'
+  upload_sbom_release_assets:
+    description: 'Upload SBOM as a release asset'
+    required: false
+    default: 'true'
+
+runs:
+  using: 'composite'
+  steps:
+    - name: Install Cosign
+      if: ${{ inputs.install_cosign == 'true' }}
+      uses: sigstore/cosign-installer@d58896d6a1865668819e1d91763c7751a165e159 # v3.9.2
+
+    - name: Generate SBOM
+      if: ${{ inputs.generate_sbom == 'true' }}
+      uses: anchore/sbom-action@da167eac915b4e86f08b264dbdbc867b61be6f0c # v0.20.5
+      with:
+        artifact-name: sbom_${{ inputs.sbom_name }}.spdx.json
+        output-file: ./sbom_${{ inputs.sbom_name }}.spdx.json
+        image: ${{ inputs.image_tag }}
+        upload-release-assets: ${{ inputs.upload_sbom_release_assets  }}
+
+    - name: Sign Container Image
+      shell: bash
+      run: |
+        cosign sign -y "${{ inputs.image }}"
+
+    - name: Attach SBOM Attestation
+      shell: bash
+      run: |
+        image="${{ inputs.image }}"
+        sbom_file="./sbom_${{ inputs.sbom_name }}.spdx.json"
+        cosign attest -y --predicate "${sbom_file}" --type spdxjson "${image}"

--- a/.github/workflows/build-images-base.yaml
+++ b/.github/workflows/build-images-base.yaml
@@ -80,9 +80,6 @@ jobs:
         id: qemu
         uses: docker/setup-qemu-action@29109295f81e9208d7d86ff1c6c12d2833863392 # v3.6.0
 
-      - name: Install Cosign
-        uses: sigstore/cosign-installer@d58896d6a1865668819e1d91763c7751a165e159 # v3.9.2
-
       # Warning: since this is a privileged workflow, subsequent workflow job
       # steps must take care not to execute untrusted code.
       - name: Checkout pull request branch (NOT TRUSTED)
@@ -132,24 +129,13 @@ jobs:
           tags: |
             quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/cilium-runtime:${{ steps.runtime-tag.outputs.tag }}
 
-      - name: Sign Container Image Runtime
+      - name: Generate SBOM, Sign, and Attest Runtime Image
         if: ${{ steps.cilium-runtime-tag-in-repositories.outputs.exists == 'false' }}
-        run: |
-          cosign sign -y quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/cilium-runtime@${{ steps.docker_build_release_runtime.outputs.digest }}
-
-
-      - name: Generate SBOM
-        if: ${{ steps.cilium-runtime-tag-in-repositories.outputs.exists == 'false' }}
-        uses: anchore/sbom-action@da167eac915b4e86f08b264dbdbc867b61be6f0c # v0.20.5
+        uses: ./.github/actions/cosign
         with:
-          artifact-name: sbom_cilium-runtime_${{ steps.runtime-tag.outputs.tag }}.spdx.json
-          output-file: ./sbom_cilium-runtime_${{ steps.runtime-tag.outputs.tag }}.spdx.json
-          image: quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/cilium-runtime:${{ steps.runtime-tag.outputs.tag }}
-
-      - name: Attach SBOM attestation to container image
-        if: ${{ steps.cilium-runtime-tag-in-repositories.outputs.exists == 'false' }}
-        run: |
-          cosign attest -y --predicate sbom_cilium-runtime_${{ steps.runtime-tag.outputs.tag }}.spdx.json --type spdxjson quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/cilium-runtime@${{ steps.docker_build_release_runtime.outputs.digest }}
+          image: "quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/cilium-runtime@${{ steps.docker_build_release_runtime.outputs.digest }}"
+          image_tag: "quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/cilium-runtime:${{ steps.runtime-tag.outputs.tag }}"
+          sbom_name: "cilium-runtime_${{ steps.runtime-tag.outputs.tag }}"
 
       - name: Image Release Digest Runtime
         if: ${{ steps.cilium-runtime-tag-in-repositories.outputs.exists == 'false' }}
@@ -221,23 +207,13 @@ jobs:
           tags: |
             quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/cilium-builder:${{ steps.builder-tag.outputs.tag }}
 
-      - name: Sign Container Image Builder
+      - name: Generate SBOM, Sign, and Attest Builder Image
         if: ${{ steps.cilium-builder-tag-in-repositories.outputs.exists == 'false' }}
-        run: |
-          cosign sign -y quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/cilium-builder@${{ steps.docker_build_release_builder.outputs.digest }}
-
-      - name: Generate SBOM
-        if: ${{ steps.cilium-builder-tag-in-repositories.outputs.exists == 'false' }}
-        uses: anchore/sbom-action@da167eac915b4e86f08b264dbdbc867b61be6f0c # v0.20.5
+        uses: ./.github/actions/cosign
         with:
-          artifact-name: sbom_cilium-builder_${{ steps.builder-tag.outputs.tag }}.spdx.json
-          output-file: ./sbom_cilium-builder_${{ steps.builder-tag.outputs.tag }}.spdx.json
-          image: quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/cilium-builder:${{ steps.builder-tag.outputs.tag }}
-
-      - name: Attach SBOM attestation to container image
-        if: ${{ steps.cilium-runtime-tag-in-repositories.outputs.exists == 'false' }}
-        run: |
-          cosign attest -y --predicate sbom_cilium-builder_${{ steps.builder-tag.outputs.tag }}.spdx.json --type spdxjson quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/cilium-builder@${{ steps.docker_build_release_builder.outputs.digest }}
+          image: "quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/cilium-builder@${{ steps.docker_build_release_builder.outputs.digest }}"
+          image_tag: "quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/cilium-builder:${{ steps.builder-tag.outputs.tag }}"
+          sbom_name: "cilium-builder_${{ steps.builder-tag.outputs.tag }}"
 
       - name: Image Release Digest Builder
         if: ${{ steps.cilium-builder-tag-in-repositories.outputs.exists == 'false' }}

--- a/.github/workflows/build-images-beta.yaml
+++ b/.github/workflows/build-images-beta.yaml
@@ -119,23 +119,12 @@ jobs:
           build-args: |
             OPERATOR_VARIANT=${{ matrix.name }}
 
-      - name: Install Cosign
-        uses: sigstore/cosign-installer@d58896d6a1865668819e1d91763c7751a165e159 # v3.9.2
-
-      - name: Sign Container Image
-        run: |
-          cosign sign -y quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-${{ github.event.inputs.suffix }}@${{ steps.docker_build_release.outputs.digest }}
-
-      - name: Generate SBOM
-        uses: anchore/sbom-action@da167eac915b4e86f08b264dbdbc867b61be6f0c # v0.20.5
+      - name: Generate SBOM, Sign Images, and Attach Attestations
+        uses: ./.github/actions/cosign
         with:
-          artifact-name: sbom_${{ matrix.name }}_${{ github.event.inputs.tag }}.spdx.json
-          output-file: ./sbom_${{ matrix.name }}_${{ github.event.inputs.tag }}.spdx.json
-          image: quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-${{ github.event.inputs.suffix }}:${{ github.event.inputs.tag }}
-
-      - name: Attach SBOM attestation to container image
-        run: |
-          cosign attest -y --predicate sbom_${{ matrix.name }}_${{ github.event.inputs.tag }}.spdx.json --type spdxjson quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-${{ github.event.inputs.suffix }}@${{ steps.docker_build_release.outputs.digest }}
+          image: "quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-${{ github.event.inputs.suffix }}@${{ steps.docker_build_release.outputs.digest }}"
+          image_tag: "quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-${{ github.event.inputs.suffix }}:${{ github.event.inputs.tag }}"
+          sbom_name: "${{ matrix.name }}_${{ github.event.inputs.tag }}"
 
       - name: Image Release Digest
         shell: bash

--- a/.github/workflows/build-images-ci.yaml
+++ b/.github/workflows/build-images-ci.yaml
@@ -238,9 +238,6 @@ jobs:
             echo build="true" >> $GITHUB_OUTPUT
           fi
 
-      - name: Install Cosign
-        uses: sigstore/cosign-installer@d58896d6a1865668819e1d91763c7751a165e159 # v3.9.2
-
       - name: CI Build ${{ matrix.name }}
         uses: docker/build-push-action@263435318d21b8e681c14492fe198d362a7d2c83 # v6.18.0
         id: docker_build_ci
@@ -292,51 +289,31 @@ jobs:
             MODIFIERS="NOSTRIP=1"
             OPERATOR_VARIANT=${{ matrix.name }}
 
-      - name: Sign Container Images
-        if: ${{ steps.check.outputs.build != '' }}
-        run: |
-          cosign sign -y quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci@${{ steps.docker_build_ci.outputs.digest }}
-          if [[ "${{ steps.docker_build_ci_detect_race_condition.outcome }}" != 'skipped' ]]; then
-            cosign sign -y quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci@${{ steps.docker_build_ci_detect_race_condition.outputs.digest }}
-          fi
-          if [[ "${{ steps.docker_build_ci_unstripped.outcome }}" != 'skipped' ]]; then
-            cosign sign -y quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci@${{ steps.docker_build_ci_unstripped.outputs.digest }}
-          fi
-
-      - name: Generate SBOM
-        if: ${{ matrix.name != 'cilium-cli' }}
-        uses: anchore/sbom-action@da167eac915b4e86f08b264dbdbc867b61be6f0c # v0.20.5
+      - name: Generate SBOM, Sign, and Attest Main CI Image
+        if: ${{ matrix.name != 'cilium-cli' && steps.check.outputs.build != '' }}
+        uses: ./.github/actions/cosign
         with:
-          artifact-name: sbom_ci_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx.json
-          output-file: ./sbom_ci_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx.json
-          image: quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci:${{ steps.tag.outputs.tag }}
+          image: "quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci@${{ steps.docker_build_ci.outputs.digest }}"
+          image_tag: "quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci:${{ steps.tag.outputs.tag }}"
+          sbom_name: "ci_${{ matrix.name }}_${{ steps.tag.outputs.tag }}"
 
-      - name: Generate SBOM (race)
+      - name: Generate SBOM, Sign, and Attest Race Detection Image
         if: ${{ matrix.name != 'cilium-cli' && steps.docker_build_ci_detect_race_condition.outcome != 'skipped' }}
-        uses: anchore/sbom-action@da167eac915b4e86f08b264dbdbc867b61be6f0c # v0.20.5
+        uses: ./.github/actions/cosign
         with:
-          artifact-name: sbom_ci_race_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx.json
-          output-file: ./sbom_ci_race_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx.json
-          image: quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci:${{ steps.tag.outputs.tag }}-race
+          image: "quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci@${{ steps.docker_build_ci_detect_race_condition.outputs.digest }}"
+          image_tag: "quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci:${{ steps.tag.outputs.tag }}-race"
+          sbom_name: "ci_race_${{ matrix.name }}_${{ steps.tag.outputs.tag }}"
+          install_cosign: 'false'
 
-      - name: Generate SBOM (unstripped)
+      - name: Generate SBOM, Sign, and Attest Unstripped Image
         if: ${{ matrix.name != 'cilium-cli' && steps.docker_build_ci_unstripped.outcome != 'skipped' }}
-        uses: anchore/sbom-action@da167eac915b4e86f08b264dbdbc867b61be6f0c # v0.20.5
+        uses: ./.github/actions/cosign
         with:
-          artifact-name: sbom_ci_unstripped_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx.json
-          output-file: ./sbom_ci_unstripped_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx.json
-          image: quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci:${{ steps.tag.outputs.tag }}-unstripped
-
-      - name: Attach SBOM attestation to container image
-        if: ${{ matrix.name != 'cilium-cli' }}
-        run: |
-          cosign attest -y --predicate sbom_ci_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx.json --type spdxjson quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci@${{ steps.docker_build_ci.outputs.digest }}
-          if [[ "${{ steps.docker_build_ci_detect_race_condition.outcome }}" != 'skipped' ]]; then
-            cosign attest -y --predicate sbom_ci_race_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx.json --type spdxjson quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci@${{ steps.docker_build_ci_detect_race_condition.outputs.digest }}
-          fi
-          if [[ "${{ steps.docker_build_ci_unstripped.outcome }}" != 'skipped' ]]; then
-            cosign attest -y --predicate sbom_ci_unstripped_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx.json --type spdxjson quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci@${{ steps.docker_build_ci_unstripped.outputs.digest }}
-          fi
+          image: "quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci@${{ steps.docker_build_ci_unstripped.outputs.digest }}"
+          image_tag: "quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci:${{ steps.tag.outputs.tag }}-unstripped"
+          sbom_name: "ci_unstripped_${{ matrix.name }}_${{ steps.tag.outputs.tag }}"
+          install_cosign: 'false'
 
       - name: CI Image Releases digests
         shell: bash

--- a/.github/workflows/build-images-hotfixes.yaml
+++ b/.github/workflows/build-images-hotfixes.yaml
@@ -125,25 +125,20 @@ jobs:
             CILIUM_RUNTIME_IMAGE=${{ env.CILIUM_RUNTIME_IMAGE }}
             OPERATOR_VARIANT=${{ matrix.name }}
 
-      - name: Install Cosign
-        uses: sigstore/cosign-installer@d58896d6a1865668819e1d91763c7751a165e159 # v3.9.2
-
-      - name: Sign Container Image
-        run: |
-          cosign sign -y quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-dev@${{ steps.docker_build_release.outputs.digest }}
-          cosign sign -y quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci@${{ steps.docker_build_release.outputs.digest }}
-
-      - name: Generate SBOM
-        uses: anchore/sbom-action@da167eac915b4e86f08b264dbdbc867b61be6f0c # v0.20.5
+      - name: Generate SBOM, Sign Images, and Attach Attestations
+        uses: ./.github/actions/cosign
         with:
-          artifact-name: sbom_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx.json
-          output-file: ./sbom_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx.json
-          image: quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-dev:${{ steps.tag.outputs.tag }}
+          image: "quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-dev@${{ steps.docker_build_release.outputs.digest }}"
+          image_tag: "quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-dev:${{ steps.tag.outputs.tag }}"
+          sbom_name: "${{ matrix.name }}_${{ steps.tag.outputs.tag }}"
 
-      - name: Attach SBOM attestation to container image
-        run: |
-          cosign attest -y --predicate sbom_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx.json --type spdxjson quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-dev@${{ steps.docker_build_release.outputs.digest }}
-          cosign attest -y --predicate sbom_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx.json --type spdxjson quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci@${{ steps.docker_build_release.outputs.digest }}
+      - name: Generate SBOM, Sign Images, and Attach Attestations
+        uses: ./.github/actions/cosign
+        with:
+          image: "quay.io/${{ env.QUAY_ORGANIZATION_DEV }}/${{ matrix.name }}-ci@${{ steps.docker_build_release.outputs.digest }}"
+          sbom_name: "${{ matrix.name }}_${{ steps.tag.outputs.tag }}"
+          install_cosign: 'false'
+          generate_sbom: 'false'
 
       - name: Image Release Digest
         shell: bash

--- a/.github/workflows/build-images-releases.yaml
+++ b/.github/workflows/build-images-releases.yaml
@@ -123,30 +123,23 @@ jobs:
             CILIUM_RUNTIME_IMAGE=${{ env.CILIUM_RUNTIME_IMAGE }}
             OPERATOR_VARIANT=${{ matrix.name }}
 
-      - name: Install Cosign
-        uses: sigstore/cosign-installer@d58896d6a1865668819e1d91763c7751a165e159 # v3.9.2
-
-      - name: Sign Container Image
-        run: |
-          if [ "${{ env.PUSH_TO_DOCKER_HUB }}" == "true" ]; then
-            cosign sign -y docker.io/${{ github.repository_owner }}/${{ matrix.name }}@${{ steps.docker_build_release.outputs.digest }}
-          fi
-          cosign sign -y quay.io/${{ env.QUAY_ORGANIZATION }}/${{ matrix.name }}@${{ steps.docker_build_release.outputs.digest }}
-
-      - name: Generate SBOM
-        uses: anchore/sbom-action@da167eac915b4e86f08b264dbdbc867b61be6f0c # v0.20.5
+      - name: Generate SBOM, Sign Images, and Attach Attestations
+        uses: ./.github/actions/cosign
         with:
-          artifact-name: sbom_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx.json
-          output-file: ./sbom_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx.json
-          image: quay.io/${{ env.QUAY_ORGANIZATION }}/${{ matrix.name }}:${{ steps.tag.outputs.tag }}
-          upload-release-assets: false
+          image: "quay.io/${{ env.QUAY_ORGANIZATION }}/${{ matrix.name }}@${{ steps.docker_build_release.outputs.digest }}"
+          image_tag: "quay.io/${{ env.QUAY_ORGANIZATION }}/${{ matrix.name }}:${{ steps.tag.outputs.tag }}"
+          sbom_name: "${{ matrix.name }}_${{ steps.tag.outputs.tag }}"
+          upload_sbom_release_assets: 'false'
 
-      - name: Attach SBOM attestation to container image
-        run: |
-          if [ "${{ env.PUSH_TO_DOCKER_HUB }}" == "true" ]; then
-            cosign attest -y --predicate sbom_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx.json --type spdxjson docker.io/${{ github.repository_owner }}/${{ matrix.name }}@${{ steps.docker_build_release.outputs.digest }}
-          fi
-          cosign attest -y --predicate sbom_${{ matrix.name }}_${{ steps.tag.outputs.tag }}.spdx.json --type spdxjson quay.io/${{ env.QUAY_ORGANIZATION }}/${{ matrix.name }}@${{ steps.docker_build_release.outputs.digest }}
+      - name: Generate SBOM, Sign Images, and Attach Attestations
+        if: ${{ env.PUSH_TO_DOCKER_HUB == 'true' }}
+        uses: ./.github/actions/cosign
+        with:
+          image: "docker.io/${{ github.repository_owner }}/${{ matrix.name }}@${{ steps.docker_build_release.outputs.digest }}"
+          sbom_name: "${{ matrix.name }}_${{ steps.tag.outputs.tag }}"
+          generate_sbom: 'false'
+          install_cosign: 'false'
+          upload_sbom_release_assets: 'false'
 
       - name: Image Release Digest
         shell: bash

